synapse/replication/http/devices.py

```diff
@@ -78,5 +79,107 @@ async def _handle_request(  # type: ignore[override]
         return 200, user_devices
 
 
+class ReplicationUploadKeysForUserRestServlet(ReplicationEndpoint):
+    """Ask master to upload keys for the user and send them out over federation to
+    update other servers.
+
+    This must happen on master so that the results can be correctly cached in
+    the database and streamed to workers.( Is this accurate?)
+
+    Calls to e2e_keys_handler.upload_keys_for_user(user_id, device_id, keys) on
+    master to accomplish this.
+
+    Defined in https://spec.matrix.org/v1.4/client-server-api/#post_matrixclientv3keysupload
+    Request format(borrowed and expanded from KeyUploadServlet):
+
+        POST /_synapse/replication/upload_keys_for_user
+
+    {
+        "user_id": "<user_id>",
+        "device_id": "<device_id>",
+        "keys": {
+            "device_keys": {
+                "user_id": "<user_id>",
+                "device_id": "<device_id>",
+                "valid_until_ts": <millisecond_timestamp>,
+                "algorithms": [
+                    "m.olm.curve25519-aes-sha2",
+                ]
+                "keys": {
+                    "<algorithm>:<device_id>": "<key_base64>",
+                },
+                "signatures:" {
+                    "<user_id>" {
+                        "<algorithm>:<device_id>": "<signature_base64>"
+                    }
+                }
+            },
+            "fallback_keys": {
+                "<algorithm>:<device_id>": "<key_base64>",
+                "signed_<algorithm>:<device_id>": {
+                    "fallback": true,
+                    "key": "<key_base64>",
+                    "signatures": {
+                        "<user_id>": {
+                            "<algorithm>:<device_id>": "<key_base64>"
+                        }
+                    }
+                }
+            }
+            "one_time_keys": {
+                "<algorithm>:<key_id>": "<key_base64>"
+            },
+        }
+    }
+    Response is equivalent to ` /_matrix/client/v3/keys/upload`
+    response, e.g.:
+
+    {
+        "one_time_key_counts": {
+            "curve25519": 10,
+            "signed_curve25519": 20
+        }
+    }
```

Comment:
Yeah, it's slightly different though. Because of the passing of the information in the _serialize_json() includes having the `user_id` and the `device_id` which isn't part of the same request as the original prepended to the beginning of the request...(I think)

Reply:
Took me a minute, but I see what you mean. How about f2c780b?